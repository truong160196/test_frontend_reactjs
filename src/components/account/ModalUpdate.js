import React from 'react';
import { connect } from 'react-redux';

import './account.css';

import * as Types from '../../constant/ActionTypes';

import {
    actUploadImage,
    actPostUpdate,
} from '../../actions/account/index';

class ModalUpdate extends React.Component {
    constructor(props) {
        super(props);
        this.formRef = React.createRef();
        this.fileUpload = React.createRef();
        this.state = props.account.model;
    }

    componentDidMount = async() => {
        //
    }

    componentWillReceiveProps = (props) => {
        const {account} = props;

        const {imageAvatar} = this.state;

        if (account && account.image !== imageAvatar) {
            this.setState({
                imageAvatar: account.image
            })
        }
    }


    closeModal = () => {
        const { closeModalUpdate } = this.props;

        closeModalUpdate(false);
    }

    handleChange = (event) => {
        event.preventDefault();
        const { name, value } = event.target

        const dataState = [];

        dataState[name] = {
            value: value,
            error: ''
        }

        const validEmailRegex = RegExp(['^[0-9]+$']);

        if (name === "phoneStore") {
            const checkRegex = validEmailRegex.test(value);

            if (checkRegex === false && value) {
                dataState[name].error = 'Phone is number type'
            } else {
                dataState[name].error = ''
            }
        }

        this.setState(dataState);
    }

    handleSubmit = (event) => {
        event.preventDefault();

        const {postUpdate} = this.props;

        const {
            idStore,
            idRedInvoice,
            nameStore,
            addressStore,
            districtStore,
            cityStore,
            phoneStore,
            nameRedInvoice,
            addressRedInvoice,
            districtRedInvoice,
            cityRedInvoice,
            taxRedInvoice,
            imageAvatar,
        } = this.state;

        const request = {
            idStore,
            idRedInvoice,
            nameStore: nameStore.value,
            addressStore: addressStore.value,
            districtStore: districtStore.value,
            cityStore: cityStore.value,
            phoneStore: phoneStore.value,
            nameRedInvoice: nameRedInvoice.value,
            addressRedInvoice: addressRedInvoice.value,
            districtRedInvoice: districtRedInvoice.value,
            cityRedInvoice: cityRedInvoice.value,
            taxRedInvoice: taxRedInvoice.value,
            imageAvatar: imageAvatar,
        }

        postUpdate(request);
    }

    handleSelectImage() {
        this.fileUpload.current.click();
    }

    upLoadImage = () => {
        const { uploadImage } = this.props;

        const files = this.fileUpload.current.files;
        
        if (files && files.length > 0) {
            const request = {
                file: files[0]
            }

            this.fileUpload.current.value = null;
            uploadImage(request);
        }

    }

    render() {
        const {
            isOpenModal
        } = this.props

        const {
            nameStore,
            addressStore,
            districtStore,
            cityStore,
            phoneStore,
            nameRedInvoice,
            addressRedInvoice,
            districtRedInvoice,
            cityRedInvoice,
            taxRedInvoice,
            imageAvatar,
        } = this.state;

        const selectDistrict = Types.DISTRICT.map((item, index) => {
            return (<option key={index.toString()} value={item}>{item}</option>)
        })

        const selectCity = Types.CITY.map((item, index) => {
            return (<option key={index.toString()} value={item}>{item}</option>)
        })

        return (
            <div
                className={`modal modal-update-user ${isOpenModal === true ? 'show-modal' : ''}`}
                id="modal-user" tabindex="-1"
            >
                <div className="modal-dialog" role="document">
                    <div className="modal-content">
                        <div className="modal-header">
                            <i class="fas fa-edit"></i>
                            <h5 className="modal-title">EDIT STORE PROFILE</h5>
                            <button
                                type="button"
                                className="close"
                                data-dismiss="modal"
                                aria-label="Close"
                                onClick={this.closeModal}
                            >
                                <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body row">
                            <div className="col-sm-12 col-lg-4" >
                                <h4>STORE IMAGE</h4>
                                <div
                                    className="avatar"
                                    onClick={() => this.handleSelectImage()}
                                >
                                    <img
                                        src={imageAvatar}
                                    />
                                </div>
                                <input
                                    type="file"
                                    id="upload"
                                    accept="image/png, image/jpeg, image/jpg"
                                    ref={this.fileUpload}
                                />
                                <div className="btn-group">
                                    <button type="button" className="btn btn-secondary btn-remove">Remove</button>
                                    <button
                                        type="button"
                                        className="btn btn-secondary btn-upload"
                                        onClick={this.upLoadImage}
                                    >
                                        Upload Image
                                    </button>
                                </div>
                            </div>
                            <div className="col-sm-12 col-lg-8" >
                                <form
                                    className="form-update"
                                    ref={this.formRef}
                                >
                                    <h4>BASIC INFO</h4>
                                    <div className="form-group">
                                        <label>Store Name</label>
                                        <input
                                            type="text"
                                            name="nameStore"
                                            value={nameStore.value}
                                            className="form-control"
                                            placeholder="Store Name"
                                            onChange={this.handleChange}
                                        />
                                    </div>
                                    <div className="form-group">
                                        <label>Store Address</label>
                                        <div className="row">
                                            <div className="col-sm-12 col-lg-6">
                                                <input
                                                    type="text"
                                                    name="addressStore"
                                                    value={addressStore.value}
                                                    className="form-control"
                                                    placeholder="Address"
                                                    onChange={this.handleChange}
                                                />
                                            </div>
                                            <div className="col-sm-12 col-lg-3">
                                                <select
                                                    className="form-control"
                                                    name="districtStore"
                                                    value={districtStore.value}
                                                    onChange={this.handleChange}
                                                >
                                                <option>District</option>
                                                {selectDistrict}
                                                </select>
                                            </div>
                                            <div className="col-sm-12 col-lg-3">
                                                <select
                                                    className="form-control"
                                                    name="cityStore"
                                                    value={cityStore.value}
                                                    onChange={this.handleChange}
                                                >
                                                    <option>City</option>
                                                    {selectCity}
                                                </select>
                                            </div>
                                        </div>
                                    </div>
                                    <div className="form-group">
                                        <label>Phone #</label>
                                        <input
                                            type="text"
                                            className="form-control"
                                            placeholder="Phone"
                                            name="phoneStore"
                                            autocomplete="off"
                                            value={phoneStore.value}
                                            onChange={this.handleChange}
                                            />
                                        <p
                                            className="error"
                                            style={{
                                                display: phoneStore.error && phoneStore.error.length > 0 ? 'block' : 'none'
                                            }}
                                        >
                                            {phoneStore.error}
                                        </p>
                                    </div>
                                    <h4>RED INVOICE INFO</h4>
                                    <div className="form-group">
                                        <label>Company Name</label>
                                        <input
                                            type="text"
                                            className="form-control"
                                            placeholder="Company Name"
                                            name="nameRedInvoice"
                                            value={nameRedInvoice.value}
                                            onChange={this.handleChange}
                                            />
                                    </div>
                                    <div className="form-group">
                                        <label>Company Address</label>
                                        <div className="row">
                                            <div className="col-sm-12 col-lg-6">
                                                <input
                                                    type="text"
                                                    className="form-control"
                                                    placeholder="Address"
                                                    name="addressRedInvoice"
                                                    value={addressRedInvoice.value}
                                                    onChange={this.handleChange}
                                                />
                                            </div>
                                            <div className="col-sm-12 col-lg-3">
                                                <select
                                                    className="form-control"
                                                    name="districtRedInvoice"
                                                    value={districtRedInvoice.value}
                                                    onChange={this.handleChange}
                                                >
                                                <option>District</option>
                                                {selectDistrict}
                                                </select>
                                            </div>
                                                <div className="col-sm-12 col-lg-3">
                                                    <select
                                                        className="form-control"
                                                        name="cityRedInvoice"
                                                        value={cityRedInvoice.value}
                                                        onChange={this.handleChange}
                                                        >
                                                        <option>City</option>
                                                        {selectCity}
                                                    </select>
                                            </div>
                                        </div>
                                    </div>
                                    <div className="form-group">
                                        <label>MST</label>
                                        <input
                                            type="text"
                                            className="form-control"
                                            placeholder="MST"
                                            name="taxRedInvoice"
                                            value={taxRedInvoice.value}
                                            onChange={this.handleChange}
                                        />
                                    </div>
                                    <div className="form-group button-submit">
                                        <button
                                            className={`btn btn-secondary btn-save 
                                            ${phoneStore.error && phoneStore.error.length > 0 ? 'disabled' : ''}`}
                                            onClick={
                                                phoneStore.error && phoneStore.error.length > 0 ? () => {}
                                                : (event) => this.handleSubmit(event)}
                                        >
                                            Save
                                        </button>
                                        <button
                                            className="btn btn-secondary btn-cancel"
                                            >
                                            Cancel
                                        </button>
                                    </div>
                                </form>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = state => ({ ...state });

const mapDispatchToProps = dispatch => ({
    uploadImage: request => {
      dispatch(actUploadImage(request));
    },
    postUpdate: request => {
        dispatch(actPostUpdate(request));
      },
});

export default connect(mapStateToProps, mapDispatchToProps)(ModalUpdate);