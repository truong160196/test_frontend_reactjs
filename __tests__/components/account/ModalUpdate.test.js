/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */
import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import ModalUpdate from '../../../src/components/account/ModalUpdate';

jest.mock('react-redux', () => ({
  connect: (mapState, mapDispatch) => ReactComponent => ReactComponent,
}));
console.warn = () => {};

describe('⚫ ModalUpdate correctly', () => {
  const account = {
    user: {
      id: 1,
      name: "K.O.I The.",
      address: "521 Hồ Tùng Mậu, D1, HCM",
      district: "Ward 1",
      city: "HCM",
      phone: "3388869944",
      logoUrl: "",
      red_invoice_id: 1,
      redInvoices: {
        id: 1,
        name: "K.O.I The International Company",
        address: "9682 Wakehurst Avenue Arlington Heights, IL, 60004",
        district: "Heights",
        city: "IL",
        taxCode: "P77744944",
      },
    },
    model: {
      idStore: 0,
      idRedInvoice: 0,
      nameStore: {
          value: '',
          error: ''
      },
      addressStore: {
          value: '',
          error: ''
      },
      districtStore: {
          value: '',
          error: ''
      },
      cityStore: {
          value: '',
          error: ''
      },
      phoneStore: {
          value: '',
          error: ''
      },
      nameRedInvoice: {
          value: '',
          error: ''
      },
      addressRedInvoice: {
          value: '',
          error: ''
      },
      districtRedInvoice: {
          value: '',
          error: ''
      },
      cityRedInvoice: {
          value: '',
          error: ''
      },
      taxRedInvoice: {
          value: '',
          error: ''
      }, 
      imageAvatar: '',
    }
  }

  const uploadImageMock = jest.fn();
  const postUpdateMock = jest.fn();
  const closeModalUpdateMock = jest.fn();

  const accountComponent = () => (
    <ModalUpdate
      account={account}
      uploadImage={uploadImageMock}
      postUpdate={postUpdateMock}
      closeModalUpdate={closeModalUpdateMock}
      isOpenModal={true}
    />
  );

  let accountComponents = Renderer.create(
    accountComponent()
    ,
  );

  it('render accountComponents update begin', () => {
    expect(accountComponents.toJSON()).toMatchSnapshot();
  });

  it('test function closeModal when user click open modal', () => {
    const wrapper = shallow(accountComponent());
    wrapper.instance().closeModal();
  });

  it('test function handleChange when change value input', () => {
    const wrapper = shallow(accountComponent());
    
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: "nameStore",
        value: "Name Store",
      }
    }

    wrapper.instance().handleChange(event);

    expect(wrapper.state('nameStore').value).toEqual("Name Store");

  });

  it('test function handleChange when change value input phone have value is number type', () => {
    const wrapper = shallow(accountComponent());
    
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: "phoneStore",
        value: "121454551",
      }
    }

    wrapper.instance().handleChange(event);

    expect(wrapper.state('phoneStore').value).toEqual("121454551");
    expect(wrapper.state('phoneStore').error).toEqual("");
  });

  it('test function handleChange when change value input phone have value is text type', () => {
    const wrapper = shallow(accountComponent());
    
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: "phoneStore",
        value: "phoneStore",
      }
    }

    wrapper.instance().handleChange(event);

    expect(wrapper.state('phoneStore').value).toEqual("phoneStore");
    expect(wrapper.state('phoneStore').error).toEqual("Please enter a valid phone number");
  });

  it('test function handleSubmit when submit form', () => {
    const wrapper = shallow(accountComponent());
    
    const event = {
      preventDefault: jest.fn(),
    }

    wrapper.instance().handleSubmit(event);
  });


  it('test function handleSelectImage when select image from local', () => {
    const wrapper = shallow(accountComponent());

    wrapper.instance().fileUpload = {
      current: {
        click: jest.fn(),
      }
    };

    wrapper.instance().handleSelectImage();
  });

  it('test function upLoadImage when upload image', () => {
    const wrapper = shallow(accountComponent());

    wrapper.instance().fileUpload = {
      current: {
        click: jest.fn(),
        files: [
          {fileName: 'example.jpg'}
        ]
      }
    };
    
    wrapper.instance().upLoadImage();
  });

  it('test function upLoadImage when upload image when have not file', () => {
    const wrapper = shallow(accountComponent());

    wrapper.instance().fileUpload = {
      current: {
        click: jest.fn(),
        files: []
      }
    };
    
    wrapper.instance().upLoadImage();
  });


  it('test function componentWillReceiveProps when have image data', () => {
    const wrapper = shallow(accountComponent());
    
    wrapper.instance().setState({
      imageAvatar: null,
    });

    const props = {
      account: {
        image: 'http://example.jpg',
      }
    }

    wrapper.instance().componentWillReceiveProps(props);

    expect(wrapper.state('imageAvatar')).toEqual("http://example.jpg");
  });

  it('test function componentWillReceiveProps when have not image data', () => {
    const wrapper = shallow(accountComponent());
    
    wrapper.instance().setState({
      imageAvatar: "http://example.jpg",
    });

    const props = {
      account: {
        image: 'http://example.jpg',
      }
    }

    wrapper.instance().componentWillReceiveProps(props);

    expect(wrapper.state('imageAvatar')).toEqual("http://example.jpg");
  });
});
