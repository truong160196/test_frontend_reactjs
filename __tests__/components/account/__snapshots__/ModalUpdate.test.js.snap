// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`⚫ ModalUpdate correctly render accountComponents update begin 1`] = `
<div
  className="modal modal-update-user show-modal"
  id="modal-user"
  tabindex="-1"
>
  <div
    className="modal-dialog"
    role="document"
  >
    <div
      className="modal-content"
    >
      <div
        className="modal-header"
      >
        <i
          class="fas fa-edit"
        />
        <h5
          className="modal-title"
        >
          EDIT STORE PROFILE
        </h5>
        <button
          aria-label="Close"
          className="close"
          data-dismiss="modal"
          onClick={[Function]}
          type="button"
        >
          <span
            aria-hidden="true"
          >
            ×
          </span>
        </button>
      </div>
      <div
        className="modal-body row"
      >
        <div
          className="col-sm-12 col-lg-4"
        >
          <h4>
            STORE IMAGE
          </h4>
          <div
            className="avatar"
            onClick={[Function]}
          >
            <img
              src=""
            />
          </div>
          <input
            accept="image/png, image/jpeg, image/jpg"
            id="upload"
            type="file"
          />
          <div
            className="btn-group"
          >
            <button
              className="btn btn-secondary btn-remove"
              type="button"
            >
              Remove
            </button>
            <button
              className="btn btn-secondary btn-upload"
              onClick={[Function]}
              type="button"
            >
              Upload Image
            </button>
          </div>
        </div>
        <div
          className="col-sm-12 col-lg-8"
        >
          <form
            className="form-update"
          >
            <h4>
              BASIC INFO
            </h4>
            <div
              className="form-group"
            >
              <label>
                Store Name
              </label>
              <input
                className="form-control"
                name="nameStore"
                onChange={[Function]}
                placeholder="Store Name"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label>
                Store Address
              </label>
              <div
                className="row"
              >
                <div
                  className="col-sm-12 col-lg-6"
                >
                  <input
                    className="form-control"
                    name="addressStore"
                    onChange={[Function]}
                    placeholder="Address"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="col-sm-12 col-lg-3"
                >
                  <select
                    className="form-control"
                    name="districtStore"
                    onChange={[Function]}
                    value=""
                  >
                    <option>
                      District
                    </option>
                    <option
                      value="Ward 1"
                    >
                      Ward 1
                    </option>
                    <option
                      value="Ward 2"
                    >
                      Ward 2
                    </option>
                    <option
                      value="Ward 3"
                    >
                      Ward 3
                    </option>
                    <option
                      value="Ward 4"
                    >
                      Ward 4
                    </option>
                    <option
                      value="Ward 5"
                    >
                      Ward 5
                    </option>
                    <option
                      value="Ward 6"
                    >
                      Ward 6
                    </option>
                    <option
                      value="Ward 7"
                    >
                      Ward 7
                    </option>
                    <option
                      value="Ward 8"
                    >
                      Ward 8
                    </option>
                    <option
                      value="Ward 9"
                    >
                      Ward 9
                    </option>
                    <option
                      value="Ward 10"
                    >
                      Ward 10
                    </option>
                    <option
                      value="Ward 11"
                    >
                      Ward 11
                    </option>
                    <option
                      value="Ward 12"
                    >
                      Ward 12
                    </option>
                    <option
                      value="Bình Tân Ward"
                    >
                      Bình Tân Ward
                    </option>
                    <option
                      value="Bình Thạnh Ward"
                    >
                      Bình Thạnh Ward
                    </option>
                    <option
                      value="Gò Vấp Ward"
                    >
                      Gò Vấp Ward
                    </option>
                    <option
                      value="Phú Nhuận Ward"
                    >
                      Phú Nhuận Ward
                    </option>
                    <option
                      value="Tân Bình Ward"
                    >
                      Tân Bình Ward
                    </option>
                    <option
                      value="Tân Phú Ward"
                    >
                      Tân Phú Ward
                    </option>
                    <option
                      value="Thủ Đức Ward"
                    >
                      Thủ Đức Ward
                    </option>
                  </select>
                </div>
                <div
                  className="col-sm-12 col-lg-3"
                >
                  <select
                    className="form-control"
                    name="cityStore"
                    onChange={[Function]}
                    value=""
                  >
                    <option>
                      City
                    </option>
                    <option
                      value="Ho Chi Minh"
                    >
                      Ho Chi Minh
                    </option>
                  </select>
                </div>
              </div>
            </div>
            <div
              className="form-group"
            >
              <label>
                Phone #
              </label>
              <input
                autocomplete="off"
                className="form-control"
                name="phoneStore"
                onChange={[Function]}
                placeholder="Phone"
                type="text"
                value=""
              />
              <p
                className="error"
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                
              </p>
            </div>
            <h4>
              RED INVOICE INFO
            </h4>
            <div
              className="form-group"
            >
              <label>
                Company Name
              </label>
              <input
                className="form-control"
                name="nameRedInvoice"
                onChange={[Function]}
                placeholder="Company Name"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group"
            >
              <label>
                Company Address
              </label>
              <div
                className="row"
              >
                <div
                  className="col-sm-12 col-lg-6"
                >
                  <input
                    className="form-control"
                    name="addressRedInvoice"
                    onChange={[Function]}
                    placeholder="Address"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="col-sm-12 col-lg-3"
                >
                  <select
                    className="form-control"
                    name="districtRedInvoice"
                    onChange={[Function]}
                    value=""
                  >
                    <option>
                      District
                    </option>
                    <option
                      value="Ward 1"
                    >
                      Ward 1
                    </option>
                    <option
                      value="Ward 2"
                    >
                      Ward 2
                    </option>
                    <option
                      value="Ward 3"
                    >
                      Ward 3
                    </option>
                    <option
                      value="Ward 4"
                    >
                      Ward 4
                    </option>
                    <option
                      value="Ward 5"
                    >
                      Ward 5
                    </option>
                    <option
                      value="Ward 6"
                    >
                      Ward 6
                    </option>
                    <option
                      value="Ward 7"
                    >
                      Ward 7
                    </option>
                    <option
                      value="Ward 8"
                    >
                      Ward 8
                    </option>
                    <option
                      value="Ward 9"
                    >
                      Ward 9
                    </option>
                    <option
                      value="Ward 10"
                    >
                      Ward 10
                    </option>
                    <option
                      value="Ward 11"
                    >
                      Ward 11
                    </option>
                    <option
                      value="Ward 12"
                    >
                      Ward 12
                    </option>
                    <option
                      value="Bình Tân Ward"
                    >
                      Bình Tân Ward
                    </option>
                    <option
                      value="Bình Thạnh Ward"
                    >
                      Bình Thạnh Ward
                    </option>
                    <option
                      value="Gò Vấp Ward"
                    >
                      Gò Vấp Ward
                    </option>
                    <option
                      value="Phú Nhuận Ward"
                    >
                      Phú Nhuận Ward
                    </option>
                    <option
                      value="Tân Bình Ward"
                    >
                      Tân Bình Ward
                    </option>
                    <option
                      value="Tân Phú Ward"
                    >
                      Tân Phú Ward
                    </option>
                    <option
                      value="Thủ Đức Ward"
                    >
                      Thủ Đức Ward
                    </option>
                  </select>
                </div>
                <div
                  className="col-sm-12 col-lg-3"
                >
                  <select
                    className="form-control"
                    name="cityRedInvoice"
                    onChange={[Function]}
                    value=""
                  >
                    <option>
                      City
                    </option>
                    <option
                      value="Ho Chi Minh"
                    >
                      Ho Chi Minh
                    </option>
                  </select>
                </div>
              </div>
            </div>
            <div
              className="form-group"
            >
              <label>
                MST
              </label>
              <input
                className="form-control"
                name="taxRedInvoice"
                onChange={[Function]}
                placeholder="MST"
                type="text"
                value=""
              />
            </div>
            <div
              className="form-group button-submit"
            >
              <button
                className="btn btn-secondary btn-save 
                                            "
                onClick={[Function]}
              >
                Save
              </button>
              <button
                className="btn btn-secondary btn-cancel"
              >
                Cancel
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
